src/lib.rs

```diff
@@ -1,564 +1,570 @@
-                record.n_blocks.iter().fold(0, |acc, blk| acc + blk.length) as usize;
-            soft_masks_count += record
-                .soft_mask_blocks
-                .iter()
-                .fold(0, |acc, blk| acc + blk.length) as usize;
-        }
-        Ok(TwoBitFileInfo {
-            file_size: metadata(&self.path)?.len(),
-            chromosomes: self.sequences.len(),
-            total_sequence_length: total_length,
-            hard_masks_count,
-            soft_masks_count,
+    fn from_value_reader(mut reader: ValueReader<R>) -> Result<Self> {
+        let sequences = reader.sequence_records()?;
+        Ok(Self {
+            reader,
+            sequences,
+            softmask_enabled: false,
         })
     }
 
-    /// Get all hard blocks (N-blocks) of a chromosome
+    /// Reads a partial sequence of a chromosome.
     ///
-    /// * `chr` Name of the chromosome from the 2bit file
-    pub fn full_hard_masked_blocks(&self, chr: &str) -> Result<Vec<Block>, Error> {
-        match self.sequence_record(chr) {
-            Ok(record) => Ok(record.n_blocks),
-            Err(e) => Err(e),
-        }
-    }
+    /// * `chr` – name of the chromosome
+    /// * `range` – selected range of the sequence (`..` for full sequence)
+    pub fn read_sequence(
+        &mut self,
+        chr: impl AsRef<str>,
+        range: impl RangeBounds<usize>,
+    ) -> Result<String> {
+        const NUC: &[u8; 4] = b"TCAG";
+
+        let seq = self.sequences.query(chr)?;
+        let reader = &mut self.reader;
+
+        let start = match range.start_bound() {
+            Bound::Included(&v) => v,
+            Bound::Excluded(&v) => v + 1,
+            Bound::Unbounded => 0,
+        };
+        let end = match range.end_bound() {
+            Bound::Included(&v) => v + 1,
+            Bound::Excluded(&v) => v,
+            Bound::Unbounded => seq.length,
+        };
 
-    /// Get hard blocks (N-blocks) of a region on a chromosome
-    ///
-    /// * `chr` Name of the chromosome from the 2bit file
-    pub fn hard_masked_blocks(
-        &self,
-        chr: &str,
-        start: usize,
-        end: usize,
-    ) -> Result<Vec<Block>, Error> {
-        match self.sequence_record(chr) {
-            Ok(record) => {
-                let mut result = Vec::new();
-                for block in record.n_blocks {
-                    let block_end = block.start + block.length;
-                    if block_end as usize <= start {
-                        continue;
-                    } else if block.start as usize > end {
-                        break;
-                    }
-                    result.push(block)
-                }
-                Ok(result)
-            }
-            Err(e) => Err(e),
+        let first_byte = start / 4;
+        reader.seek(SeekFrom::Start(seq.offset))?; // beginning of the DNA sequence
+        reader.seek(SeekFrom::Current(first_byte as _))?; // position where we want to start reading
```

Comment:
A typecast to underscore? Is this type erasure like a `(void*)` pointer in C?

Reply:
No, it's not type erasure - it's basically when the type of the destination is known, you don't have to type it explicitly when using `as`, and can just use an underscore. E.g. `let x: i64 = 10_i32 as _`;